### PR TITLE
Better extension point scanning

### PR DESCRIPTION
**Pull request by thomaslevesque**
As discussed in chat with @blairconrad

WIP because I still need to implement "light loading" of assemblies for TFMs that don't have `ReflectionOnlyLoad`.

### CODE REVIEW

**Review by blairconrad — CHANGES_REQUESTED** [review]
As usual, very nice. I like the unwrapping of the LINQ.
My biggest concern is the docs, but I have some additional minor questions and comments.

File: docs/scanning-for-extension-points.md [hunk 1]
```diff
@@ -19,9 +19,13 @@ Please see their individual documentation to learn how each of these is used.
 On startup, FakeItEasy searches:
 
 * its own assembly,
-* assemblies already loaded in the current AppDomain and
+* assemblies already loaded in the current AppDomain,
+* assemblies referenced by already loaded assemblies and
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply docs/scanning-for-extension-points.md:23]: > assemblies already loaded in the current AppDomain that reference FakeItEasy

Yes, I thought about that, but it was already true before, and the docs didn't mention it. Anyway, assemblies that don't reference FakeItEasy can't implement extension points.

> My concern is that there's a currently-loaded assembly that _does not itself reference FIE, but that references an assembly that does and that contains an extension point class_.
> It seems unlikely to happen, but I think it's possible, and if so, we will not find that class. I'm okay not finding it, but figure we should at least write it down.

That's a good point. I'll clarify the docs

File: src/FakeItEasy/Core/TypeCatalogue.cs [hunk 2]
```diff
@@ -68,77 +68,113 @@ public IEnumerable<Type> GetAvailableTypes()
...
 
-#if FEATURE_REFLECTIONONLYLOAD
-                // A reflection-only loaded assembly can't be scanned for types, so fully load it before returning it.
-                try
+                foreach (var referencedAssemblyName in assembly.GetReferencedAssemblies())
                 {
-                    assembly = Assembly.Load(assembly.GetName());
+                    if (TryLoadAssemblyIfItReferencesFakeItEasy(referencedAssemblyName, out var referencedAssembly))
+                    {
+                        assembliesToScan.Add(referencedAssembly);
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Core/TypeCatalogue.cs:93]: Good point

> or equivalently enrolling line 93 into the if on line 94

I think that's the best approach

File: src/FakeItEasy/Core/TypeCatalogue.cs [hunk 3]
```diff
@@ -68,77 +68,113 @@ public IEnumerable<Type> GetAvailableTypes()
...
+                WarnFailedToLoadAssembly(file, ex);
+                return false;
+            }
+
+            if (!assembly.ReferencesFakeItEasy())
+            {
+                return false;
+            }
+
+            return true;
```
[the remark this answers is not in the document]
Comment (thomaslevesque) [reply src/FakeItEasy/Core/TypeCatalogue.cs:148]: I like it. In fact, I probably would have written it like this, but that code was initially doing other things, which I removed when I got rid of `ReflectionOnlyLoad`, and I forgot to refactor.

**Comment by thomaslevesque** [comment]
Pushed fixup commits with the suggested changes

File: src/FakeItEasy/Core/TypeCatalogue.cs (hunk 2 above)
[the remark this answers is not in the document]
Comment (blairconrad) [reply src/FakeItEasy/Core/TypeCatalogue.cs]: Me too.

**Review by blairconrad — APPROVED** [review]
Wunderbar. Thanks!

**Comment by thomaslevesque** [comment]
Squashed

**Review by blairconrad — APPROVED** [review]

**Comment by blairconrad** [comment]
Thanks, @thomaslevesque! I know of at least one loyal user that will be thrilled with the change.

**Comment by thomaslevesque** [comment]
Thanks for the suggestions and merge!

> I know of at least one loyal user that will be thrilled with the change.

Oh? I wonder who 🤔

**Comment by blairconrad** [comment]
This change has also been released as part of [FakeItEasy 7.0.0](https://github.com/FakeItEasy/FakeItEasy/releases/tag/7.0.0).

**Pull request merged**